### PR TITLE
Verify backend integrity on import

### PR DESCRIPTION
**Pull request by nkoep**
~~This is still a WIP but I wanted to record the progress here already anyway.~~

This PR lays the groundwork for testing backend functions in isolation. It mainly consists of two parts:
1. The backend module is now loaded dynamically using python import hooks (see https://www.python.org/dev/peps/pep-0302/). In short, it allows to intercept import instructions and allows to control what is returned to the runtime for code like `import geomstats.backend as gs`. This finally gets rid of *-imports.
1. When a backend is loaded, the importer now checks at import-time whether all required functions are provided by a backend. The single source of truth which defines the backend interface is therefore the ~~`BACKEND_FUNCTIONS`~~ `BACKEND_ATTRIBUTES` variable in `geomstats/_backend/__init__.py`. ~~Right now, the importer only checks the integrity. I'm thinking of changing this and essentially creating the returned module from scratch to ensure nothing that isn't specified in the `BACKEND_ATTRIBUTES` mapping ends up being exposed.~~ If a backend cannot currently support a certain function/feature, the function should still be provided and simply raise a `NotImplementedError` for now.

In short, the `BACKEND_ATTRIBUTES` variable is used to enforce that every backend exposes the same functions.

### CODE REVIEW

**Comment by nkoep** [comment]
Alright, the PR should be ready for review now, @ninamiolane.

File: geomstats/_backend/tensorflow/__init__.py [hunk 1]
```diff
@@ -74,15 +141,15 @@ def get_mask_float(indices, mask_shape):
...
 
-def duplicate_array(x, n_samples, axis=0):
+def _duplicate_array(x, n_samples, axis=0):
     multiples = _np.ones(ndim(x) + 1, dtype=_np.int32)
     multiples[axis] = n_samples
     return tile(to_ndarray(x, ndim(x) + 1), multiples)
 
 
-def get_vectorized_mask_float(
+def _get_vectorized_mask_float(
```
[the remark this answers is not in the document]
Comment (ninamiolane) [reply geomstats/_backend/tensorflow/__init__.py:150]: Maybe coordinate with @pchauchat on this? :/

File: geomstats/_backend/tensorflow/linalg.py [hunk 2]
```diff
@@ -13,6 +13,10 @@
 sqrtm = tf.linalg.sqrtm
 
 
+def eig(*args, **kwargs):
```
Comment (ninamiolane) [inline geomstats/_backend/tensorflow/linalg.py:16]: Do you want to use _raise_not_implemented_error here?

File: geomstats/_backend/tensorflow/__init__.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (pchauchat) [reply geomstats/_backend/tensorflow/__init__.py:150]: We discussed it yesterday, I had some issues with Git so I had to keep my renaming. From what I understood, the important part was to add the underscore, since the function is to be used only internally. Then, whether we call it (vectorized_)mask_from_indices or get_(vectorized_)mask_float is another question.

File: geomstats/_backend/tensorflow/linalg.py (hunk 2 above)
Comment (nkoep) [reply geomstats/_backend/tensorflow/linalg.py:16]: For the moment, yes, because the pytorch backend didn't provide an `eig` implementation so far. Most of the functions which currently raise a `NotImplementedError` I plan on implementing (where possible) in the course of the next PR adding unit tests for backend functions.

File: geomstats/_backend/numpy/__init__.py [hunk 3]
```diff
@@ -94,30 +86,8 @@
...
-def byte_to_float(x):
-    return x
-
-
-def any(x, axis=0):
-    return _np.any(x, axis)
-
 
+# XXX(nkoep): Can we get rid of this now?
 def while_loop(cond, body, loop_vars, maximum_iterations):
```
[the remark this answers is not in the document]
Comment (nkoep) [reply geomstats/_backend/numpy/__init__.py:91]: Alright, good to know. I'll look into this at a later point then.

File: geomstats/_backend/__init__.py [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+                    raise RuntimeError(error) from None
+                else:
+                    setattr(new_submodule, attribute_name, attribute)
+
+        from numpy import pi
+        new_module.pi = pi
+
+        return new_module
+
+    def find_module(self, fullname, path=None):
```
[the remark this answers is not in the document]
Comment (nkoep) [reply geomstats/_backend/__init__.py:191]: We don't use/need it, but it has to be specified for the finder to adhere to the protocol specified in https://www.python.org/dev/peps/pep-0302/.

File: geomstats/_backend/numpy/__init__.py [hunk 5]
```diff
@@ -167,20 +137,13 @@ def vectorize(x, pyfunc, multiple_args=False, signature=None, **kwargs):
     return _np.vectorize(pyfunc, signature=signature)(x)
 
 
+# XXX(nkoep): Can we get rid of this now?
```
[the remark this answers is not in the document]
Comment (nkoep) [reply geomstats/_backend/numpy/__init__.py:140]: I often use it to annotate code with questions I want to answer later, which I think is a quite common use. I can remove it if you prefer (or change it into a TODO).

File: geomstats/_backend/numpy/linalg.py [hunk 6]
```diff
@@ -9,22 +9,21 @@
...
 
-def is_symmetric(x, tol=TOL):
-    new_x = to_ndarray(x, to_ndim=3)
+def _is_symmetric(x, tol=_TOL):
+    new_x = _to_ndarray(x, to_ndim=3)
     return (np.abs(new_x - np.transpose(new_x, axes=(0, 2, 1))) < tol).all()
 
 
-def expsym(x):
+def _expsym(x):
```
[the remark this answers is not in the document]
Comment (nkoep) [reply geomstats/_backend/numpy/linalg.py:26]: Yes, I mentioned this to @pchauchat as well. I think it's a good idea to mark internal functions with an underscore to make it easier to understand what is part of the compatibility layer and what isn't.

File: geomstats/_backend/pytorch/__init__.py [hunk 7]
```diff
@@ -433,50 +389,16 @@ def outer(*args, **kwargs):
...
-    return torch.cov(*args, **kwargs)
-
-
 def eval(x):
     return x
 
 
 def ndim(x):
+    if not torch.is_tensor(x):
+        x = torch.tensor(x)
```
[the remark this answers is not in the document]
Comment (nkoep) [reply geomstats/_backend/pytorch/__init__.py:398]: Jepp, definitely agree. Type errors should just trigger unhandled cause they indicate bugs in the code which should be fixed!

File: geomstats/_backend/pytorch/__init__.py [hunk 8]
```diff
@@ -573,6 +471,6 @@ def array_from_sparse(indices, data, target_shape):
 
 def from_vector_to_diagonal_matrix(x):
```
[the remark this answers is not in the document]
Comment (nkoep) [reply geomstats/_backend/pytorch/__init__.py:552]: The function is used in the `SymmetricMatrices` class so it needs to be exposed. The question is whether it's actually needed somewhere else or if it could be turned into a method of the class instead.

File: geomstats/_backend/numpy/__init__.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ninamiolane) [reply geomstats/_backend/numpy/__init__.py:217]: Let's keep it then. I also just found this https://github.com/JohnPostlethwait/fixme,which could be useful for us to grab these later (and make them into github issues).

File: geomstats/_backend/numpy/linalg.py [hunk 9]
```diff
@@ -9,22 +9,21 @@
...
+# TODO(nina): Clean this import.
+from geomstats._backend.numpy.__init__ import to_ndarray as _to_ndarray
 
-TOL = 1e-10
+_TOL = 1e-10
 
 
-def is_symmetric(x, tol=TOL):
-    new_x = to_ndarray(x, to_ndim=3)
+def _is_symmetric(x, tol=_TOL):
```
Comment (ninamiolane) [inline geomstats/_backend/numpy/linalg.py:21]: This function is also here: https://github.com/geomstats/geomstats/blob/1e88292766310bbccf75bd0a9297b4a8da87f567/geomstats/geometry/matrices.py#L100

However, I think that it is fine to keep both, so that we do not have to instantiate a Matrices objects in the backend?

File: geomstats/_backend/pytorch/__init__.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (ninamiolane) [reply geomstats/_backend/pytorch/__init__.py:552]: I think this function was implemented because the np.diag function was not vectorized as we needed. Let's keep it here then. Thanks!

File: geomstats/_backend/tensorflow/__init__.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ninamiolane) [reply geomstats/_backend/tensorflow/__init__.py]: Sure! Right now in the code base it is `mask_from_indices`, so let's do `_mask_from_indices`? Unless you have strong preference for one of the two other namings?

**Comment by ninamiolane** [comment]
Great, thank you! 🎉Can't wait to merge this. It is good to go when:
- flake8 & travis pass, 
- the naming of the function `_mask_from_indices` is decided (see comment),
- see if some deep source issues can be quickly fixed: https://github.com/geomstats/geomstats/pull/515/checks?check_run_id=565113736.

(We will use deep source more and more for automatic code reviews, I'm trying to configure it these days, to see which errors we should keep and which we should ignore).

**Comment by ninamiolane** [comment]
Example of errors ignored: https://deepsource.io/gh/geomstats/geomstats/issue/FLK-D103/ignore-rules

File: geomstats/_backend/numpy/linalg.py (hunk 9 above)
Comment (nkoep) [reply geomstats/_backend/numpy/linalg.py:21]: Jepp, definitely agree. The backend should have no knowledge of the `Matrices` class.
